kernel-rs/src/virtio/virtio_disk.rs

```diff
@@ -352,6 +328,136 @@ impl VirtioDisk {
+                    barray[i].as_mut().unwrap(),
+                    true,
+                );
+
+                let mut this = guard.get_pin_mut().project();
+                // Tell the device the first index in our chain of descriptors.
+                let write_idx = (this.avail.idx as usize + num_block_written) % NUM;
+                this.avail.ring[write_idx] = descs[0].idx as _;
+
+                darray[i] = Some(descs);
+                prev_blk_idx = i;
+                num_block_written += 1;
+            }
+        }
+
+        assert!(
+            Self::finalize_write_seq(
+                guard,
+                &mut darray[next_write..=prev_blk_idx],
+                &mut barray[next_write..=prev_blk_idx],
+                num_block_written,
+                ctx,
+            ),
+            "could not finalize the last set of buffers"
+        );
+    }
+
+    // This method writes the buffers in the given range of the given array to the disk.
+    fn finalize_write_seq(
+        guard: &mut SleepableLockGuard<'_, Self>,
+        darray: &mut [Option<[Descriptor; 3]>],
+        barray: &mut [Option<Buf>],
+        num_block_written: usize,
+        ctx: &KernelCtx<'_, '_>,
+    ) -> bool {
+        let length = darray.len();
+        // Other than the arrays' length, the caller must make sure that they are corresponding with each other.
+        assert_eq!(length, barray.len());
+
+        // Assert that the chain contains at least one buffer to be written.
+        if num_block_written == 0 {
+            return false;
+        }
+
+        let mut this = guard.get_pin_mut().project();
+
+        fence(Ordering::SeqCst);
+
+        // Tell the device `num_block_written` avail ring entries are available.
+        this.avail.idx += num_block_written as u16;
+
+        fence(Ordering::SeqCst);
+
+        // SAFETY: all the three descriptors' fields are well set.
+        // Value is queue number.
+        unsafe {
+            MmioRegs::notify_queue(0);
+        }
+
+        // Wait for the disk to finish writing all the requests
+        // The request is complete when the last block is written.
+        // TODO: This may require that the `VIRTIO_F_IN_ORDER` feature has been negotiated.
+        // Another way is to check all Bufs if their disk fields are set to false.
+        barray
+            .last()
+            .unwrap()
+            .as_ref()
+            .unwrap()
+            .vdisk_request_waitchannel
+            .sleep(guard, ctx);
+
+        // Cleanup
+        // Free the descriptors
+        for dopt in darray {
+            if let Some(d) = dopt.take() {
+                guard.get_pin_mut().project().info.project().inflight[d[0].idx].b = ptr::null_mut();
+                IntoIter::new(d).for_each(|desc| guard.get_pin_mut().free(desc));
+            }
+        }
+
+        guard.wakeup(ctx.kernel());
+
+        return true;
+    }
+
     pub fn intr(self: Pin<&mut Self>, kernel: KernelRef<'_, '_>) {
```

Comment:
I suggest you change the `VirtioDisk::intr` as the following, so that we wakeup only one waitchannel.
```Rust
    pub fn intr(self: Pin<&mut Self>, kernel: KernelRef<'_, '_>) {
        // The device won't raise another interrupt until we tell it
        // we've seen this interrupt, which the following line does.
        // This may race with the device writing new entries to
        // the "used" ring, in which case we may process the new
        // completion entries in this interrupt, and have nothing to do
        // in the next interrupt, which is harmless.
        MmioRegs::intr_ack_all();

        fence(Ordering::SeqCst);

        // The device increments disk.used->idx when it
        // adds an entry to the used ring.

        let this = self.project();
        let info = this.info.project();

        let buf = loop {
            fence(Ordering::SeqCst);
            let id = this.used.ring[(*info.used_idx as usize) % NUM].id as usize;

            assert!(!info.inflight[id].status, "Disk::intr status");            

            *info.used_idx += 1;
            if *info.used_idx == this.used.id {
                // SAFETY: from the invariant, b refers to a valid
                // buffer unless it is null.
                break unsafe { &mut *info.inflight[id].b };
            }
        };
        // disk is done with buf
        buf.deref_inner_mut().disk = false;
        buf.vdisk_request_waitchannel.wakeup(kernel);
    }
}
```

Reply:
What is the purpose of waking up only one waitchannel? I tried running usertests using this code, and it seemed to incur deadlock in `concreate` with `CPUS=1` and in `truncate3`, which is the point where the segment is first committed, with `CPUS=3`.

Reply:
> What is the purpose of waking up only one waitchannel?

It's because we sleep only for a single waitchannel in `VirtioDisk::finalize_write_seq` but wake up multiple waitchannels in `VirtioDisk::intr`. Since waking up waitchannels are costly, I wanted to change the code so that we only wakeup the last one.

> I tried running usertests using this code, and it seemed to incur deadlock in concreate with CPUS=1 and in truncate3, which is the point where the segment is first committed, with CPUS=3.

* Maybe this code doesn't work well when working with multiple processes.
* Are you sure its a deadlock? It seems like it's just because we didn't wake up the proper waitchannel(s).